### PR TITLE
For SM120, the MoE NVFP4 kernel of B12X is added to SGLang

### PR DESCRIPTION
**Pull request by hsr1234563**
## Motivation

When running Qwen3.5-397B NVFP4 (https://huggingface.co/nvidia/Qwen3.5-397B-A17B-NVFP4) on SM120, we found that the MoE component accounts for an extremely large proportion of runtime, nearly 50%, which requires optimization. The test data for this PR is as follows:

## Modifications

Directly enable the MoE NVFP4 kernel of b12x (https://github.com/lukealonso/b12x/tree/master) within the MoE backend of SGLang.
Currently, this PR uses flashinfer_cutedsl as the backend name, yet it actually links against b12x directly. A new commit will replace the calling interfaces here with the corresponding APIs from FlashInfer.

Follow-up: make sure the FlashInfer B12X W4A16 path (NVFP4 weights with bf16 activations) works in SGLang too. See #33710.

## Accuracy Tests

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/shao/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/shao/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">

  |   | 1 | 2
-- | -- | -- | --
397B-MoE-b12x | nvfp4 | 0.884 | 0.8838
397B-MoE-triton | nvfp4 | 0.899 | 0.874
397B-MoE-triton | FP8 | 0.889 | 0.884

</body>

</html>

## Speed Tests and Profiling

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/shao/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/shao/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">
<meta charset=utf-8>

<div data-scroll-padding=4 data-visibility=hover data-direction=horizontal
style='--tw-space-y-reverse: 0;margin-block: 0px 16px;--scrollbar-size: 6px;
--scrollbar-inset: 0px;--scrollbar-thumb-top-offset: 6px;--scroll-area-scroll-padding: 4px;
display:grid;grid-template: 1fr / 1fr;border-color:color(srgb 0.894118 0.894118 0.894118 / 0.0737255);
border-image: none 100% / 1 / 0 stretch;border-radius: 6px;color:rgba(228, 228, 228, 0.92);
font-variant-ligatures: normal;font-variant-caps: normal;orphans: 2;text-align:
start;widows: 2;-webkit-text-stroke-width: 0px;text-decoration-thickness: initial;
text-decoration-style: initial;text-decoration-color: initial'>

MoE Backend | EP | ISL | OSL | Concurrency | TTFT (ms) | TPOT (ms) | QPS | Output tok/s/user | Total tok/s | QPS/GPU | Output tok/s/GPU | Total tok/s/GPU
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
triton | 1 | 4096 | 1024 | 4 | 4443.95 | 22.65 | 0.14 | 37.07 | 741.45 | 0.0181 | 18.54 | 92.68
flashinfer_cutlass | 1 | 4096 | 1024 | 4 | 4422.22 | 19.36 | 0.17 | 42.26 | 845.15 | 0.0206 | 21.13 | 105.64
flashinfer_cutedsl_sm120 (b12x) | 1 | 4096 | 1024 | 4 | 4282.32 | 18.18 | 0.17 | 44.74 | 894.79 | 0.0218 | 22.37 | 111.85

</body>

</html>

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.
Latest PR Test (AMD ROCm 10): :heavy_minus_sign: **No AMD PR run found for this commit**.

### CODE REVIEW

**Comment by nvpohanh** [comment]
cc @b8zhong

File: python/sglang/srt/layers/moe/moe_runner/flashinfer_cutedsl_sm120.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+
+def ensure_cutedsl_sm120_runtime(layer: torch.nn.Module) -> None:
+    if getattr(layer, "_cutedsl_sm120_ready", False):
+        return
+
+    if not torch.cuda.is_available():
+        raise RuntimeError("flashinfer_cutedsl_sm120 requires CUDA.")
+
+    major, _ = torch.cuda.get_device_capability(layer.w13_weight.device)
```
Comment (b8zhong) [inline python/sglang/srt/layers/moe/moe_runner/flashinfer_cutedsl_sm120.py:55]: Use is_sm120_supported

**Review by b8zhong — COMMENTED** [review]
Do not call it flashinfer_cutedsl. Let's call it flashinfer_b12x

File: python/sglang/srt/layers/moe/moe_runner/flashinfer_cutedsl_sm120.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    return FlashinferCombineInput(hidden_states=output)
+
+
+def fused_experts_deepep_to_flashinfer_cutedsl_sm120_fp4(
+    dispatch_output: DeepEPLLDispatchOutput,
+    quant_info: CuteDslSm120Fp4MoeQuantInfo,
+    runner_config: MoeRunnerConfig,
+) -> DeepEPLLCombineInput:
+    raise NotImplementedError(
+        "flashinfer_cutedsl_sm120 does not support deepep path yet. "
```
Comment (b8zhong) [inline python/sglang/srt/layers/moe/moe_runner/flashinfer_cutedsl_sm120.py:257]: I don't think A2A works on SM120. Do not create this registration if not

File: python/sglang/srt/layers/moe/moe_runner/flashinfer_cutedsl_sm120.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        "Use a2a=none/flashinfer."
+    )
+
+
+def _maybe_register_sm120_fused_funcs() -> None:
```
Comment (b8zhong) [inline python/sglang/srt/layers/moe/moe_runner/flashinfer_cutedsl_sm120.py:262]: Do not name it maybe

File: python/sglang/srt/layers/moe/moe_runner/flashinfer_cutedsl_sm120.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        topk_ids=topk_ids,
+        activation="silu",
+        quant_mode="nvfp4",
+        source_format="modelopt_nvfp4",
+        input_scales_are_reciprocal=True,
+        input_scales_static=True,
+        apply_router_weight_on_input=False,
+        output=output,
+    )
+    from b12x.integration import TPMoEScratchCaps, plan_tp_moe_scratch
```
Comment (b8zhong) [inline python/sglang/srt/layers/moe/moe_runner/flashinfer_cutedsl_sm120.py:142]: Let's just import the Flashinfer one, since Flashinfer will be upgraded soon

**Review by b8zhong — COMMENTED** [review]
When the quantization is ModelOpt FP4, please set it automatically in server args hook for maybe Qwen35 if it can be better for all num tokens

**Comment by hsr1234563** [comment]
Thanks @b8zhong. Do you want me to continue pushing this PR? Because this PR is has been pending for a long time. I do not know whether I should keep working on it? Also, if you think this function is meaningful now, I will continue to work on it.

**Comment by hsr1234563** [comment]
Or is this function in someone else's PR? Pls tell me, Thanks
